### PR TITLE
fix(net): avoid deadlock on stayin_alive calls

### PR DESCRIPTION
**Pull request by dignifiedquire**
## Change checklist

- [ ] Self-review.
- [ ] Documentation updates if relevant.
- [ ] Tests if relevant.

### CODE REVIEW

**Comment by rklaehn** [comment]
I can confirm that this fixes the immediate issue with the pings deadlocking. On main I get a timeout connecting to a peer that has a huge initial peer map, on this branch it works.

**Comment by rklaehn** [comment]
Maybe the "throttling" due to sending pings to yourself leads to something going slower, which reduces packet loss. I wonder how my unbounded everything PR would fare.

**Comment by dignifiedquire** [comment]
Ready for rereview, tests are back to passing now.

**Comment by rklaehn** [comment]
So, tried this with a really big peers.postcard file and it seems to work just fine.

There is still a pretty big bottleneck when requesting from multiple peers, which I think is due to a bottleneck in the actor.

```
❯ ls -l idd/peers.postcard 
-rw-------  1 rklaehn  staff  9390 Sep 28 15:33 idd/peers.postcard

❯ ntimesll.sh 5 cargo run --release get --ticket ebiwlve7jj6flt3zoakty2nzvbydfk5fkdc7vvtsquemqzhxtpcheaicaiafm67fulcfoagavaayrrcxjqcticlmcocymflb4dzn4rwnrnxhglhck3pplnwhpq2uedubwz6qaai
Transferred 957.40 MiB in 30 seconds, 32.23 MiB/s
Transferred 957.40 MiB in 33 seconds, 28.61 MiB/s
Transferred 957.40 MiB in 33 seconds, 29.00 MiB/s
Transferred 957.40 MiB in 34 seconds, 28.48 MiB/s
Transferred 957.40 MiB in 33 seconds, 28.92 MiB/s

❯ ntimesll.sh 10 cargo run --release get --ticket ebiwlve7jj6flt3zoakty2nzvbydfk5fkdc7vvtsquemqzhxtpcheaicaiafm67fulcfoagavaayrrcxjqcticlmcocymflb4dzn4rwnrnxhglhck3pplnwhpq2uedubwz6qaai
Transferred 957.40 MiB in 67 seconds, 14.36 MiB/s
Transferred 957.40 MiB in 68 seconds, 13.98 MiB/s
Transferred 957.40 MiB in 70 seconds, 13.59 MiB/s
Transferred 957.40 MiB in 72 seconds, 13.23 MiB/s
Transferred 957.40 MiB in 73 seconds, 13.16 MiB/s
Transferred 957.40 MiB in 73 seconds, 13.20 MiB/s
Transferred 957.40 MiB in 75 seconds, 12.84 MiB/s
Transferred 957.40 MiB in 75 seconds, 12.72 MiB/s
Transferred 957.40 MiB in 75 seconds, 12.72 MiB/s
Transferred 957.40 MiB in 76 seconds, 12.62 MiB/s

❯ ntimesll.sh 10 cargo run --release get --ticket ebiwlve7jj6flt3zoakty2nzvbydfk5fkdc7vvtsquemqzhxtpcheaicaiafm67fulcfoagavaayrrcxjqcticlmcocymflb4dzn4rwnrnxhglhck3pplnwhpq2uedubwz6qaai
Transferred 957.40 MiB in 67 seconds, 14.24 MiB/s
Transferred 957.40 MiB in 70 seconds, 13.72 MiB/s
Transferred 957.40 MiB in 70 seconds, 13.63 MiB/s
Transferred 957.40 MiB in 71 seconds, 13.47 MiB/s
Transferred 957.40 MiB in 71 seconds, 13.55 MiB/s
Transferred 957.40 MiB in 71 seconds, 13.45 MiB/s
Transferred 957.40 MiB in 71 seconds, 13.46 MiB/s
Transferred 957.40 MiB in 71 seconds, 13.45 MiB/s
Transferred 957.40 MiB in 73 seconds, 13.15 MiB/s
Transferred 957.40 MiB in 73 seconds, 13.15 MiB/s
```

File: iroh-net/src/magicsock/endpoint.rs [hunk 1]
```diff
@@ -312,21 +325,25 @@ impl Endpoint {
         let now = Instant::now();
         let (udp_addr, derp_region, _should_ping) = self.addr_for_send(&now);
         if let Some(derp_region) = derp_region {
-            self.start_ping(SendAddr::Derp(derp_region), now, DiscoPingPurpose::Cli)
-                .await;
+            if let Some(msg) = self.start_ping(SendAddr::Derp(derp_region), DiscoPingPurpose::Cli) {
+                todo!()
```
Comment (rklaehn) [inline iroh-net/src/magicsock/endpoint.rs:329]: ?

**Review by rklaehn — COMMENTED** [review]
I like the approach, and this seems to work fine. But there are several `todo!()` in the code.

File: iroh-net/src/magicsock/endpoint.rs (hunk 1 above)
Comment (dignifiedquire) [reply iroh-net/src/magicsock/endpoint.rs]: fixed

**Comment by dignifiedquire** [comment]
> There is still a pretty big bottleneck when requesting from multiple peers, which I think is due to a bottleneck in the actor.

Can you file a new issue for this please?

**Comment by rklaehn** [comment]
> > There is still a pretty big bottleneck when requesting from multiple peers, which I think is due to a bottleneck in the actor.
> 
> Can you file a new issue for this please?

I think it is basically this: https://github.com/n0-computer/iroh/issues/1475

I am working on getting some reliable stats in transfer_collection. I think I want to try this and then add some more details / analysis to ^...

**Comment by Arqu** [comment]
Curious, what was the crux of the issue in the end. History was re-written so bit hard to follow through.
